nixpkgs_review/nix.py

```diff
@@ -38,6 +42,61 @@ def was_build(self) -> bool:
     def is_test(self) -> bool:
         return self.name.startswith("nixosTests")
 
+    def log(self, tail: int = -1, strip_colors: bool = False) -> Optional[str]:
+        def get_log(path: Optional[str]) -> Optional[str]:
+            if path is None:
+                return None
+            system = subprocess.run(
+                ["nix", "--experimental-features", "nix-command", "log", path],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                text=True,
+                errors="backslashreplace",
+            )
+            stdout = system.stdout
+            if tail > 0 and len(stdout) > tail:
+                stdout = "This file has been truncated\n" + stdout[-tail:]
+
+            if strip_colors:
+                return strip_ansi_colors(stdout)
+            return stdout
+
+        if self.drv_path is None:
+            return None
+
+        value = get_log(self.drv_path) or get_log(self.path) or ""
+        if self.build_err_msg is not None:
+            value = "\n".join([value, self.build_err_msg])
+
+        return value
+
+    def log_path(self) -> Optional[str]:
+        if self.drv_path is None:
+            return None
+        base = os.path.basename(self.drv_path)
+        prefix = "/nix/var/log/nix/drvs/"
+        full = os.path.join(prefix, base[:2], base[2:] + ".bz2")
+        if os.path.exists(full):
+            return full
+        return None
+
+    def build_time(self) -> Optional[timedelta]:
+        log_path = self.log_path()
+        if log_path is None:
+            return None
+
+        proc = subprocess.run(
```

Comment:
This is ugly, but unfortunately I'm not aware of a better way to get the build time associated with each drv.

1. The idea to use stat on the log file is from: https://github.com/NixOS/nix/issues/1710#issuecomment-470898243
2. The reason for calling the coreutils `stat` binary rather than using the python os.stat function is that cpython does not call the new statx(2) syscall on linux, which is the only way to get the birth time of the file. Note that is not available on all filesystems either. I have a python interface to statx(2) [here](https://github.com/rmcgibbo/statx), but I figured it was easier to use coreutils stat(1) since my statx wrapper is not in nixpkgs and there's been little movement on https://bugs.python.org/issue39533

Reply:
Has this been tested on macOS?

Reply:
You can also use `ctypes` to access statx from libc instead. But than it would be also not portable to macOS since this syscall is os-specific.

Reply:
The darwin equivalent is ``/usr/bin/stat -f "%m" $file``

```
/usr/bin/stat -f "%m" /nix/var/log/nix/drvs/00/074sx96jwrvrj1ypxmzsj2g5k2fsvg-ocaml4.08.1-jackline-unstable-2020-09-03.drv.bz2
1604335341
```

but we can use the stat from coreutils

```
stat --format "%W %Y" /nix/var/log/nix/drvs/00/074sx96jwrvrj1ypxmzsj2g5k2fsvg-ocaml4.08.1-jackline-unstable-2020-09-03.drv.bz2
1604335336 1604335341
```

Reply:
I linked above to the python wrapper for statx that I have. It uses ctypes. https://github.com/rmcgibbo/statx/blob/master/statx.py